### PR TITLE
[LPT] Reduce*Transformation

### PR DESCRIPTION
**Pull request by v-Golubev**
Issue #45289

CPU Validation:
• Accuracy: developer-services/job/accuracy/1522/ (reference: developer-services/job/accuracy/1506/)
• Performance: developer-services/job/performance/777/ (reference: developer-services/job/performance/759/)

GPU Validation:
• Accuracy: developer-services/job/accuracy/1589/ (reference: developer-services/job/accuracy/1536/)

### CODE REVIEW

**Review by GlebKazantaev — APPROVED** [review]
Approved nGraph related changes

**Review by eshoguli — COMMENTED** [review]
Negative scales check are required in handling dequantization operations for any minimum & maximum operations. Please do it. Thanks!

File: inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+namespace low_precision {
+
+class TRANSFORMATIONS_API ReduceBaseTransformation : public LayerTransformation {
+public:
+    ReduceBaseTransformation(const Params& params);
+    bool transform(TransformationContext& context, ngraph::pattern::Matcher& m) const override;
+    bool canBeTransformed(const TransformationContext& context, std::shared_ptr<Node> reduce) const override;
+
+protected:
+    std::shared_ptr<opset1::Constant> RemoveDimsOfDequantizationConstant(
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp]: Method has been removed

File: inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+    // updatePrecision depends on type and parameters of the reduce
+    const bool updatePrecision = getUpdatePrecision(reduce);
+    moveDequantizationAfter(context, reduce, dequantization, updatePrecision);
+    return true;
+}
+
+bool ReduceBaseTransformation::canBeTransformed(const TransformationContext& context, std::shared_ptr<Node> reduce) const {
+    const auto dequantization = NetworkHelper::getDequantization(reduce);
+    if (dequantization.multiply == nullptr) {
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp]: Thanks, fixed!

File: inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp:38]: Negative scales check added to `ReduceMaxTransformation `and `ReduceMeanTransformation`. Tests added. Thanks!

File: inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        const std::shared_ptr<opset1::Constant>& constant,
+        std::vector<size_t> reducedAxes) const;
+    virtual void changeDequantizationValues(
+        FakeQuantizeDequantization& dequantization,
+        const std::shared_ptr<Node>& reduce) const;
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp]: Thanks, done!

File: inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+std::shared_ptr<opset1::Constant> ReduceBaseTransformation::RemoveDimsOfDequantizationConstant(
+    const std::shared_ptr<opset1::Constant>& constant,
+    std::vector<size_t> reducedAxes) const {
+    const auto constantShape = constant->get_shape();
+    if (ngraph::shape_size(constantShape) == 1ul) {
+        return NetworkHelper::toScalar(constant);
+    } else {
+        Shape reducedConstantShape(constantShape);
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp]: Thanks, done! `removeDimsOfDequantizationConstant` has been removed and used fold for dequantization constants

File: inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        // sort in descending order for easy removal of constant shape elements
+        std::sort(reducedAxes.begin(), reducedAxes.end(), std::greater<uint64_t>());
+        for (const auto& axis : reducedAxes) {
+            reducedConstantShape.erase(reducedConstantShape.begin() + axis);
+        }
+        return opset1::Constant::create(constant->get_element_type(), reducedConstantShape, constant->cast_vector<float>());
+    }
+}
+
+void ReduceBaseTransformation::changeDequantizationValues(
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/reduce_base_transformation.cpp:75]: Done, thanks! Used fold for dequantization operations. `getKeepDims` has been removed

File: inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/reduce_function.cpp [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            return std::make_shared<ngraph::op::TypeRelaxed<ngraph::opset1::ReduceMin>>(
+                std::vector<element::Type>{ precision, constant->get_element_type() },
+                std::vector<element::Type>{ precision },
+                ngraph::op::TemporaryReplaceOutputType(parent, precision).get(),
+                ngraph::op::TemporaryReplaceOutputType(constant, constant->get_element_type()).get(),
+                keepDims);
+        } else {
+            THROW_TRANSFORMATION_EXCEPTION << "unexpected Reduce type";
+        }
+    }
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/tests/ngraph_helpers/lpt_ngraph_functions/src/reduce_function.cpp]: Thanks, done!

File: inference-engine/tests/functional/inference_engine/lp_transformations/reduce_sum_transformation.cpp [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            ngraph::element::u8,
+            {},
+            ngraph::element::f32,
+            {{}, {}, {{0.1f, 1.f, 10.f}, ngraph::element::f32, {1, 3, 1, 1}}}
+        }
+    },
+    // U8: keep dims, per-channel quantization with subtract, reduction by batch
+    {
+        LayerTransformation::createParamsU8I8(),
+        {1},
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/tests/functional/inference_engine/lp_transformations/reduce_sum_transformation.cpp:73]: Fixed comment on 138 line

File: inference-engine/src/low_precision_transformations/src/reduce_mean.cpp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+bool ReduceMeanTransformation::isPrecisionPreserved(std::shared_ptr<Node> reduce) const noexcept {
+    const auto children = getChildrenRecursivelyExceptPrecisionPreserved(reduce);
+    return NetworkHelper::notAllChildrensAreFQ(children);
+}
+
+bool ReduceMeanTransformation::getUpdatePrecision(const std::shared_ptr<Node>& reduce) const {
+    const auto children = getChildrenRecursivelyExceptPrecisionPreserved(reduce);
+    if ((children.size() == 1ul) && (!this->layerTransformationsManager->isQuantized(children[0]))) {
+        return false;
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/src/reduce_mean.cpp]: WA has been removed. Now ReduceMean always has fp32 output precision

File: inference-engine/tests/functional/inference_engine/lp_transformations/reduce_mean_transformation.cpp [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+        {-2},
+        false,
+        {
+            ngraph::element::u8,
+            {{ngraph::element::f32}, {128.f}, {0.1f}}
+        },
+        {
+            ngraph::element::u8,
+            {},
+            ngraph::element::f32,
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply inference-engine/tests/functional/inference_engine/lp_transformations/reduce_mean_transformation.cpp:122]: Now ReduceMean always has fp32 output precision. This is correct behavior

File: inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <memory>
+#include <ngraph/ngraph.hpp>
+#include "layer_transformation.hpp"
+
+namespace ngraph {
+namespace pass {
+namespace low_precision {
+
+class TRANSFORMATIONS_API ReduceBaseTransformation : public LayerTransformation {
```
Comment (ilyachur) [inline inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp:15]: Could you add a documentation for transformations?

Comment (v-Golubev) [reply inference-engine/src/low_precision_transformations/include/low_precision/reduce_base_transformation.hpp:22]: Documentation added for ReduceBaseTransformation. Derived classes have the same behavior

**Review by eshoguli — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

**Pull request merged**